File: search2DMatrix.java
/*
 * Approach : finding 2d array elements using binary search algo
 * Time complexity: O(log mn)
 * Space complexity: O(1)
 */


class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int m = matrix.length;
        int n = matrix[0].length;

        int low = 0;
        int high = m * n - 1 ; //max length for the index array
        int mid = (low + (high - low))/2;

        while(low <= high) {
            mid = low + (high - low)/2; // high-low to avoid integer overload in java
            int row = mid/n;
            int col = mid%n;
            if (target == matrix[row][col]) {
                return true;
            } else if(target > matrix[row][col]) {
                low = mid + 1; //moving right
            } else {
                high = mid -1; // moving left
            }
        }
        return false;
    }
}

File: Search2DMatrix.cs
//Algo: Binary Search.
//Logically converted the array t0 1D array and then applied the BS.
//for BS converted mid to 2D matrix row and column index and the checked target,
//and accordingly reducted the search space.
//TC: O(logN)
//SC: O(1) no extra space used

public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        
        if(matrix == null){
            return false;
        }
        
        int rowSize = matrix[0].Count();
        int colSize = matrix.Count();
        
        int low = 0;
        int high = (rowSize * colSize) - 1;
        
        while(low<=high){
            int mid = low + (high-low)/2;
            int rowIndex = mid/rowSize;
            int colIndex = mid%rowSize;
            
            int midVal = matrix[rowIndex][colIndex];
            if(midVal == target){
                return true;
            }
            
            if(target < midVal){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        
        return false;
    }
}